### ISSUE TITLE
Splash screen

### ISSUE THREAD

**Issue by Grimmy347** [nodogsplash/nodogsplash#265]
I am trying to use this to make it easier for people to login to my marquee app so that it opens to the page they need to be on as soon as they login to the wireless ap but because of how nodogsplash changes iptables, it messes with the iptables I already have setup and makes it where my forwarding for my local domain name no longer works.  Is there a way to just get the splash screen without requiring the iptable changes and authentication?  That is really all I am looking for, just a way to make a page pop up after they login to the wireless AP.  I know this is not what this was intended for but it is the closest thing I have been able to find so far.  Thanks in advance!

**Comment by Grimmy347** [comment]
I am familiar with that table entry for forwarding traffic on a port to a specific ip but I am not sure how that helps my current issue.  The problem with no dog splash is when it changes the ip tables, my “webapp.local” address stops working and I have to use just the ip address of the web app.  My goal is to make the interface as dummy proof as possible so anyone can use it with little to no knowledge.  They login to the app, the popup takes them to a menu page and then when they click where they want to go, it authenticates(only doing that because I can’t seem to get around it in nodogsplash) and then takes them to the marquee page where they can type in the message they want to display.  I have all of this working now except the webapp.local address stopped working in the process after nodogsplash altered the ip tables.  I’m still learning how to navigate up tables but even if I knew how to correct what was changed, I’m pretty sure nodog will just change them again on the next restart.

**Comment by mwarning** [comment]
for webapp.local to be resolved, you may need to allow udp port 5353 in your nodogsplash configuration (along with port 53, but that should be there already by default).

**Comment by Grimmy347** [comment]
Nice, it looks like its working after I authenticate but it does not seem to work from the splash page.  If I click on the webapp.local link on the splash page it says it can not open the link because it can not find the server but if I go to my browser after that then I can open webapp.local

**Comment by bluewavenet** [comment]
@Grimmy347 
I don't understand exactly what you are trying to do or what you mean by a marquee ap, so I will make a few guesses and see if it helps.
1. You want people to connect to your wifi and immediately have your local web page pop up on their screens.
2. You don't need them to have internet access.

If this is the case you need a Forwading Authentication Service without bothering to authenticate.

**Comment by bluewavenet** [comment]
If my guess is correct, you need to do just too things, assuming your nds router has internet access itself.
1. Put a meta refresh to your web page in splash.html
2. Allow access to the web page for NDS preauth clients in the nds config

**Comment by bluewavenet** [comment]
If you don't have an internet connection for your NDS router, put Dnsmasq into offline mode.

**Comment by bluewavenet** [comment]
If you want webapp.local to resolve you will need to add it to the nds router hosts file. the default "local" domain for Dnsmasq is .lan
If the nds router is also providing dhcp to the lan (it does normally) then your local webserver can be accessed by the url: http://mywebserverhostname.lan

**Comment by Grimmy347** [comment]
Playing around with it though I also often have the message of "Tokens do not match" when clicking on the link on the splash page

**Comment by Grimmy347** [comment]
@bluewavenet 
I do not use Dnsmasq for my setup but yes, I am running a raspberry pi 3 with dhcpd running and then I setup the local domain to .local and then the hostname to webapp so I get webapp.local.  I can access my webapp.local just fine after I get past the splash page if I go to the browser and type in webapp.local but the splash page right now is doing the tokens do not match when I try to click the link on the splash page.  Before that I was getting the "Hotspot login cannot open the page..." error but playing around with the redirect setting in nodogsplash changed it from one error to the other.  I had it working earlier today but I am not sure what I did to break it.  =/  I am not sure what the meta refresh would do on the splash page.  could you explain what that would accomplish?

**Comment by bluewavenet** [comment]
I assume you are using nds v1. What operating system?
~Tokens do not match is caused by token expiry.

Can I also assume my guess about your use is correct?

Meta refresh would send people directly to your web page without stopping at the splash page.

**Comment by Grimmy347** [comment]
Yes,  I want once they finish loging into the rpi router that it will auto take them to webapp.local in the web browser.  That would be the best way for it to work.  Right now though I have the splash page pop up, they click a link which authenticates them and then takes them to the redirect link that opens in the browser.  

I am using nds v1 because I am using an older Raspbian copy to make all of the programs work together.  The tutorial I was using was older and the only way I could get it all to work was by using versions of everything from when the tutorial was written.

As far as the token issue, how would I prevent the token expiring?

I will see if I can figure out the meta refresh idea.

**Comment by Grimmy347** [comment]
Ok, so I added the meta link to go to http://webapp.local/scrolling-test/
When I login to the router it pops up with the splash page that tries to take me directly to the link but I get the following error "hotspot login cannot open the page because the server cannot be found"
I have added the firewall rule for preauth users to allow tcp port 80 to ip.of.webapp.local but that does not seem to work.

**Comment by bluewavenet** [comment]
You can see full details of how to configure this here:
https://github.com/nodogsplash/nodogsplash/tree/master/forward_authentication_service
You can just omit the nds authentication part.

**Comment by Grimmy347** [comment]
With it setup that way right now it opens the splash page and immedietely tries to redirect to the address in the meta refresh but gives the error for hoptspot login cannot open to the page because the server cannot be found.

**Comment by bluewavenet** [comment]
> Ok, so I added the meta link to go to http://webapp.local/scrolling-test/
> When I login to the router it pops up with the splash page that tries to take me directly to the link but I get the following error "hotspot login cannot open the page because the server cannot be found"

This means the client CPD cannot resolve webapp.local
Try putting in the IP address as I said earlier.

**Comment by Grimmy347** [comment]
ok, I tried that and it worked.  So what it does now is the splash page takes you to the ip but within the captive portal and any link I click on there seems to stay in the captive portal rather then opening up to a page in the web browser.  I kind of like that though.  The only problem is that once I disconnect my  raspberry ethernet cable, it no longer will do the captive portal when I login to the raspberry.  If I plug in the ethernet cable and reset then it does work.

**Comment by Grimmy347** [comment]
Thank you for all your help btw.  I have been working on this project for a while and am really close to finishing it.

**Comment by Grimmy347** [comment]
I also think its odd that the captive portal cant resolve the webapp.local but the web browser can.

**Comment by bluewavenet** [comment]
Captive Portal Detection (CPD) is a process that runs on the client device. It is what pops up the splash screen. It does it by trying to access a vendor predetermined port 80 URL that returns an expected result. If that result is not what the CPD receives then it assumes it is behind a captive portal and opens its own display window, usually a very restricted browser.
If you do not have an internet connection then the CPD will fail to resolve the special URL so it will not pop up the splash.
This is what I meant by putting Dnsmasq into offline mode. In offline mode, Dnsmasq returns its own IP address for everything.
You will need to configure a similar mode for whatever DNS server you are using.

**Comment by bluewavenet** [comment]
CPD will always use the DNS server it is given by DHCP.

**Comment by Grimmy347** [comment]
Hmmm.. I had it working without Dnsmasq last night.  all I have installed is DHCPd and hostapd and then using payara to serve the webpage

**Comment by bluewavenet** [comment]
No DNS?

**Comment by Grimmy347** [comment]
so are you saying the ip address changes in offline mode?

**Comment by Grimmy347** [comment]
I had dnsmasq installed when I was trying to do this yesterday but it seemed to make things worse so I uninstalled it

**Comment by Grimmy347** [comment]
this is my iptables right now:
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
-N ndsAUT
-N ndsNET
-N ndsRTR
-N ndsTRT
-N ndsTRU
-A INPUT -i wlan0 -j ndsRTR
-A FORWARD -i wlan0 -j ndsNET
-A FORWARD -i eth0 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i wlan0 -o eth0 -j ACCEPT
-A ndsAUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A ndsAUT -p tcp -m tcp --dport 53 -j ACCEPT
-A ndsAUT -p udp -m udp --dport 53 -j ACCEPT
-A ndsAUT -p tcp -m tcp --dport 80 -j ACCEPT
-A ndsAUT -p tcp -m tcp --dport 443 -j ACCEPT
-A ndsAUT -p tcp -m tcp --dport 22 -j ACCEPT
-A ndsAUT -p tcp -m tcp --dport 8080 -j ACCEPT
-A ndsAUT -p udp -m udp --dport 5353 -j ACCEPT
-A ndsAUT -j REJECT --reject-with icmp-port-unreachable
-A ndsNET -m mark --mark 0x100/0x700 -j DROP
-A ndsNET -m state --state INVALID -j DROP
-A ndsNET -p tcp -m tcp --tcp-flags SYN,RST SYN -j TCPMSS --clamp-mss-to-pmtu
-A ndsNET -m mark --mark 0x200/0x700 -j ACCEPT
-A ndsNET -m mark --mark 0x400/0x700 -j ndsAUT
-A ndsNET -p tcp -m tcp --dport 53 -j ACCEPT
-A ndsNET -p udp -m udp --dport 53 -j ACCEPT
-A ndsNET -p udp -m udp --dport 5353 -j ACCEPT
-A ndsNET -d 192.168.42.1/32 -p tcp -m tcp --dport 80 -j ACCEPT
-A ndsNET -j REJECT --reject-with icmp-port-unreachable
-A ndsRTR -m mark --mark 0x100/0x700 -j DROP
-A ndsRTR -m state --state INVALID -j DROP
-A ndsRTR -m state --state RELATED,ESTABLISHED -j ACCEPT
-A ndsRTR -p tcp -m tcp ! --tcp-option 2 --tcp-flags SYN SYN -j DROP
-A ndsRTR -p tcp -m tcp --dport 2050 -j ACCEPT
-A ndsRTR -m mark --mark 0x200/0x700 -j ACCEPT
-A ndsRTR -p udp -m udp --dport 53 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 53 -j ACCEPT
-A ndsRTR -p udp -m udp --dport 67 -j ACCEPT
-A ndsRTR -p udp -m udp --dport 8080 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 8080 -j ACCEPT
-A ndsRTR -p udp -m udp --dport 5353 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 22 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 80 -j ACCEPT
-A ndsRTR -p tcp -m tcp --dport 443 -j ACCEPT
-A ndsRTR -j REJECT --reject-with icmp-port-unreachable

**Comment by bluewavenet** [comment]
Just start again from scratch. Should be quite a quick process.

**Comment by Grimmy347** [comment]
Maybe if you know what you are doing.  re-flashing the sd card alone takes forever.  once I install dnsmasq I still don't quite know what I am doing so I will probably just be right back to this point once I do.  I will try again tomorrow I guess.  I know there has to be a way to get the captive portal to pull up without the Ethernet connection though since I did it last night/this morning.

**Comment by bluewavenet** [comment]
You don't need to go anywhere near iptables to make this work.
1. Install nsd on a fresh rpi.
2. edit splash.html to just have a single line containing the meta refresh.
3. add both a preauthenticated user allow tcp to the ip and port of the web server (needed in v1)
4. add the port to users to router as well
Put dnsmasq into offline mode (see dnsmasq docs)

**Comment by bluewavenet** [comment]
Leave the card reflashing and go to bed. In the morning- good to go.

**Comment by Grimmy347** [comment]
it would be that easy if I didn't have to use a very particular version of NDS/OS/WEBapp/LedMatrix.  Yeah, my brain is fried for the evening.  Thank you for your help.  I got a lot closer to what I want tonight.

**Comment by bluewavenet** [comment]
@Grimmy347 
Old versions should not be a problem.
It is much simpler than you initially thought. I use a similar configuration with the meta refresh in production and it works very well.
Then for version 1 of NDS and earlier add into nodogsplash.conf the following:
1. in FirewallRuleSet preauthenticated-users,       FirewallRule allow tcp port 80 to 192.168.1.1 (changing the ip and port as required)
2. in FirewallRuleSet users-to-router    FirewallRule allow tcp port 80 (again changing the port to match that of your webapp if required).

Get this working first then I can help with the offline mode.

**Comment by Grimmy347** [comment]
Ok, I just used a backup I had and I am at the point you mentioned above.  the meta refresh is pointing to webserver ip, the nds has the rules mentioned above setup.  Now I just need to do the offline mode with Dnsmasq

**Comment by Grimmy347** [comment]
@bluewavenet   i am not really seeing any good tutorials online on how to put dnsmasq in offline mode  =/

**Comment by bluewavenet** [comment]
@Grimmy347 
It just requires a single line addition to the dnsmasq config file ie:
address=/#/[lan_ip_address_of_NDS]

eg in my test system working here now:
address=/#/10.168.1.1

This takes effect after a restart.

You will find information about this tucked away somewhere in the dnsmasq man page at:
http://www.thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html

**Comment by Grimmy347** [comment]
I did that but it did not have the effect of making the splash page come up.

**Comment by Grimmy347** [comment]
does it matter that my webserver and nds are on the same ip?

**Comment by Grimmy347** [comment]
The moment I plug the ethernet back in to the pi then it opens the splash page but until then, nothing.

**Comment by bluewavenet** [comment]
The splash page is triggered by the device CPD trying its vendor defined portal detection url.
To do this it must receive a positive dns response from the router. If this is only working for you with an Internet connection then dnsmasq is still "online" and responds with an error with the internet feed disconnected.
Check your dmsmasq configuration
ie the line:
 address=/#/[lan_ip_address_of_NDS]
eg:
 address=/#/192.168.1.1
Make this the ONLY line in dnsmasq.conf (usually located in /etc but not sure in your case)
Reboot to make sure this has taken effect.
Test by pinging any domain name. You should get the ip of your nds box for everything.

**Comment by Grimmy347** [comment]
so the flash save that I am using uses the https://learn.adafruit.com/setting-up-a-raspberry-pi-as-a-wifi-access-point/install-software tutorial for setting up the router.  I don't know if any of those settings are what are causing the trouble.  I am thinking it could be isc-dhcp-server that it has you install causing the problem but I am not sure.

**Comment by Grimmy347** [comment]
looks like if I disable the isc-dhcp-server then when I try to connect to the rpi it says no internet connection on my phones wifi screen.  I am guessing Dnsmasq is not acting as a DHCP server due to some setting I have going on

**Comment by bluewavenet** [comment]
Sorry, I have no idea. My instructions assume you are using dnsmasq.

**Comment by bluewavenet** [comment]
Your problem is dns, not dhcp.

**Comment by Grimmy347** [comment]
I am using dnsmasq, I am just starting from where that tutorial leaves off.

**Comment by Grimmy347** [comment]
I have confirmed though that every domain I try to ping when the ethernet cable is connected, it gets the nds ip.  I can't ping it though when the cable is not connected.  I don't have an app installed to ssh atm

**Comment by Grimmy347** [comment]
service dnsmasq status I get "ignoring nameserver 192.168.42.1 - local interface"  under the line saying reading /var/run/dnsmasq/resolv.conf

I don't know if that means it will not work.  it looks like it is using my router given ip of 192.168.1.153 instead when I check it with the ethernet cable installed

**Comment by bluewavenet** [comment]
Those ip addresses will be different subnets in normal circumstances so it looks like you have a bad configuration. Your problem is about how to set up rpi and not about nds.
You would normally have the rpi based router configured with, say, 192.168.x.1 and serving dhcp in the range 192.168.x.2 to 192.168.x.254 (at maximum). 192.168.x.1 will be the default gateway as well as the DNS server address.

There is not much else I can do to help.

**Comment by Grimmy347** [comment]
Man, I am having so little luck and I think this is why I ended up using old version of everything so the tutorials actually made sense.  I backup my old work and then reflashed the sd with a fresh install of the most recent version of raspbian.  I then did apt-get update and upgrade.  I then installed NDS per your instruction but when I tried to run it, I got ip table errors "iptables: No chain/target/match by that name." and "iptables: index of deletion too big."  No clue what that means  =/

**Comment by oldtree2008** [comment]
I have the same problem.  iptables: No chain/target/match by that name.

**Comment by bluewavenet** [comment]
You might read issue #249 and perhaps ask there if someone with rpi experience can help.

**Comment by oldtree2008** [comment]
I read the issue#249 carefully.My hostapd,dhcp,dnsmasq are all ok. dnsmasq is set to offline.without nds,I can access web website through ap. my website is in router.my lan can not connect to Internet. but when I start NDS,I can not access my website.is there something wrong ?

**Comment by bluewavenet** [comment]
@oldtree2008 Yes you have something wrong :-)
As I said earlier for version 1 of NDS and earlier add into nodogsplash.conf the following:
1.   in FirewallRuleSet preauthenticated-users, add the line - FirewallRule allow tcp port 80 to 192.168.1.1 (changing the ip and port to match that of your webapp as required)
 2.  in FirewallRuleSet users-to-router, add the line - FirewallRule allow tcp port 80 (again changing the port to match that of your webapp if required).

For later versions you just need 2 above, but 1 does no harm.

**Comment by oldtree2008** [comment]
Thank you for response.My NDS version is 2.1.1. I try configure as you said. but still failed.I don't known why. is my network configure wrong ?
my Os is ubuntu 16.04.
eth0 ip is 192.168.1.102, wlan0 ip is 192.168.10.1,ap is up.
dns ： address=/#/192.168.1.102  server=/localnet/192.168.1.102
dhcp:subnet 192.168.10.0 netmask 255.255.255.0 {
  range 192.168.10.1 192.168.10.254;
  option domain-name-servers 192.168.1.102;
  option routers 192.168.10.1;
  #option broadcast-address 10.5.5.31;
  default-lease-time 600;
  max-lease-time 7200;
}
nodogsplash:
FirewallRuleSet users-to-router {
 # Nodogsplash automatically allows tcp to GatewayPort,
 # at GatewayAddress, to serve the splash page.
 # However you may want to open up other ports, e.g.
 # 53 for DNS and 67 for DHCP if the router itself is
 # providing these services.
    FirewallRule allow udp port 53
    FirewallRule allow tcp port 53
    FirewallRule allow udp port 67
 # You may want to allow ssh, http, and https to the router
 # for administration from the GatewayInterface.  If not,
 # comment these out.
   FirewallRule allow tcp port 22
   FirewallRule allow tcp port 80
   FirewallRule allow tcp port 443
}
GatewayAddress 192.168.10.1

**Comment by bluewavenet** [comment]
You do not say what hardware you are using, I assume a rpi with ethernet and wifi?
I don't use Ubuntu so not familiar with the details of setup, but it is setup that is causing the problem.
A checklist of basic requirements for this to work in your case are:
1. rpi setup as a router with ethernet interface Internet facing and wifi a separate local subnet.
2. wifi interface runs dhcp and dns services for wifi clients on the rpi. This will commonly be dnsmasq.
3. NDS bound to the wifi interface.
4. A Preauthenticated user firewall rule allowing wifi clients to access your webapp on the internet facing network (tcp 192.168.1.102:80) from the wifi network subnet (192.168.10.x).
5. A setting on the rpi DNS service to make it reply 192.168.1.102 for ANY dns enquiries received on the wifi interface subnet (192.168.10.x).

You did say earlier your webapp was running on the router and I assumed by router you meant the device NDS is running on.

**Comment by oldtree2008** [comment]
![image](https://user-images.githubusercontent.com/5414588/43946581-3522474a-9cb8-11e8-974a-e17d093fda29.png)

**Comment by oldtree2008** [comment]
I connect through wifi by my mobile. I input www.google.com ,browser display my app website.

**Comment by bluewavenet** [comment]
I'm still guessing what your actual configuration on the nds router is but it looks like it may be working.
Do you have the FirewallRuleSet preauthenticated-users setting?
This should be:
FirewallRuleSet preauthenticated-users {
    FirewallRule allow tcp port 80 to 192.168.1.102
}

**Comment by Grimmy347** [comment]
So I have updated my whole system to the latest version and I am running NDS 3.0 and when I start NDS I just get the messages:
iptables: Index of deletion too big.
iptables: No chain/target/match by that name.

I get the first one twice and the second one many times.  Any idea what would be causing that?

**Comment by Grimmy347** [comment]
Also, what guide would you recommend to use to setup ap prior to setting up NDS?

**Comment by Grimmy347** [comment]
I was not trying to install 3 but when I did the install per the guide that is what I think I got.  How can I tell for sure what is installed?

**Comment by mwarning** [comment]
@Grimmy347 what install guide did you use? Maybe not all documentation was updated properly yet.

**Comment by Grimmy347** [comment]
[syslog.txt](https://github.com/nodogsplash/nodogsplash/files/2280316/syslog.txt)
Here is my syslog after reboot with debug level 7

**Comment by Grimmy347** [comment]
https://www.raspberrypi.org/documentation/configuration/wireless/access-point.md
The first part is what I used for setting it up.  I did not use part two of this since I don't intend to have the Ethernet cable plugged in once I have everything working.

**Comment by Grimmy347** [comment]
https://pimylifeup.com/raspberry-pi-captive-portal/  is what I used for setting up NDS

**Comment by Grimmy347** [comment]
yes, the error message is when I try to run the command after startup.  It gives those two errors:
iptables: Index of deletion too big.
iptables: No chain/target/match by that name.

and NDS will not popup when I login to the wifi  =/

**Comment by mwarning** [comment]
`iptables: Too many links.` is a sign that some iptables table exists but could not have been properly removed.

**Comment by Grimmy347** [comment]
well if I flush my tables and then run nodogsplash I get:
iptables: No chain/target/match by that name.             21 times

**Comment by Grimmy347** [comment]
I tried restarting without it in rc.local and then flushing tables and running it again and I get:
iptables: No chain/target/match by that name          several times followed by 
iptables: Index of deletion too big.         followed by
iptables: No chain/target/match by that name          once... No clue what I am doing wrong here

**Comment by mwarning** [comment]
"iptables: No chain/target/match by that name." 21 times is no problem.
On NDS start, it tries do flush/remove NDS tables that might be left in case NDS has crashed.
If there are no orphaned tables left, then the error will appear (Though I thought to have it filtered, but I have an idea how to suppress those).

"iptables: Index of deletion too big." has me concerned, in this case cleanup might have failed because a table is present and could not be deleted.

**Comment by Grimmy347** [comment]
yeah, I wish I could just get this dang thing working.  I have been playing around with it for far too long now.  Most of my project wasn't too hard but this part has taken me forever.  The best I have been able to get so far is for it to do the splash screen when the ethernet cable is connected on an older version of everything but all updated I can't get NDS to work at all.

**Comment by Grimmy347** [comment]
I did it that round to make sure it was not being affected by anything else

**Comment by bluewavenet** [comment]
As I pointed out earlier, other people have this working with little problem.
You seem to be doing a lot of messing about with iptables and this is clearly confusing the issue particularly as you have not indicated what "else" you need to manipulate iptables for.
You should concentrate initially in just getting NDS working in offline mode and once achieved that start adding your additional components.

**Comment by Grimmy347** [comment]
I didn't start messing about with the iptables until I started having issues with NDS.  NDS not working is the only reason I was doing anything with them and only because it was not working.  I did a fresh install like you suggested and installed NDS/Dnsmasq and did the setting suggested with the results above.  I have now just found someone who made a image of a working captive portal AP.  I am just editing it to fit my needs.  i would have liked to figure out what I was doing wrong but I need to finish this project so I can free up time for other things.  Thanks for your attempt to help though.
